src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.cs

```diff
@@ -783,6 +637,78 @@ private Compilation CreateEmptyCompilation()
+                    InProgressState { PendingTranslationSteps: [] } => this.ProjectState,
+                    // Otherwise, reset us to whatever state the InProgressState had currently transitioned to.
+                    InProgressState inProgressState => inProgressState.PendingTranslationSteps.First().oldState,
+                    _ => throw ExceptionUtilities.UnexpectedValue(state.GetType()),
+                };
+
+                var frozenState = GetFrozenCompilationState(state);
+                Contract.ThrowIfFalse(frozenState.IsFrozen);
+                return new CompilationTracker(projectState, frozenState, this.SkeletonReferenceCache.Clone());
+
+                CompilationTrackerState GetFrozenCompilationState(CompilationTrackerState? state)
+                {
+                    if (state is FinalCompilationTrackerState finalState)
+                    {
+                        // Checked by caller.
+                        Contract.ThrowIfTrue(finalState.IsFrozen);
+                        // If we're already finalized then just return what we have, and with the frozen bit flipped so
+                        // that any future forks keep things frozen.
+                        return finalState.WithIsFrozen();
+                    }
+                    else if (state is null)
+                    {
+                        // We have no data at all. Create a frozen empty project/compilation to represent this state.
+                        var inProgressProject = this.ProjectState.RemoveAllDocuments();
+                        var compilationWithoutGeneratedDocuments = this.CreateEmptyCompilation();
+                        var compilationWithGeneratedDocuments = compilationWithoutGeneratedDocuments;
+
+                        return InProgressState.Create(
+                            isFrozen: true,
+                            compilationWithoutGeneratedDocuments,
+                            CompilationTrackerGeneratorInfo.Empty,
+                            compilationWithGeneratedDocuments,
+                            ImmutableList<(ProjectState, CompilationAndGeneratorDriverTranslationAction)>.Empty);
+                    }
+                    else if (state is InProgressState inProgressState)
+                    {
+                        // Grab whatever is in the in-progress-state so far, add any generated docs, and snap 
+                        // us to a frozen state with that information.
+                        var generatorInfo = inProgressState.GeneratorInfo;
+
+                        var compilationWithoutGeneratedDocuments = inProgressState.CompilationWithoutGeneratedDocuments;
+                        var compilationWithGeneratedDocuments = compilationWithoutGeneratedDocuments.AddSyntaxTrees(
+                            generatorInfo.Documents.States.Values.Select(state => state.GetSyntaxTree(cancellationToken)));
+
+                        return InProgressState.Create(
```

Comment:
This may be worth a comment here still saying we're using in-progress so that way the connecting of project to project and references is fixed up. Because I once again went through the mental gymnastics of "why isn't this final state?"